<a href="https://colab.research.google.com/github/JOEL1706/data-Science/blob/main/Association__rules.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
from mlxtend.frequent_patterns import apriori,association_rules
from mlxtend.preprocessing import TransactionEncoder

In [ ]:
!pip install mlxtend

In [ ]:
df=pd.read_excel("/content/Online retail.xlsx")
df.head()

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                             burgers,meatballs,eggs                                                                                                                                                                             
1                                            chutney                                                                                                                                                                             
2                                     turkey,avocado                                                                                                                                                                             
3  mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                     low fat yogurt

In [ ]:
df.tail() # checkin for the last 5 variables

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt

In [ ]:
df = df.rename(columns={'shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil': 'Transactions'})
df.head()

Transactions
0                             burgers,meatballs,eggs
1                                            chutney
2                                     turkey,avocado
3  mineral water,milk,energy bar,whole wheat rice...
4                                     low fat yogurt

In [ ]:
print(df.isnull().sum())

Transactions    0
dtype: int64


In [ ]:
df = df.iloc[:, 0].dropna().apply(lambda x: x.split(',')).tolist()

In [ ]:
trans = TransactionEncoder()
trans_df = trans.fit_transform(df)
data=pd.DataFrame(trans_df, columns=trans.columns_,dtype='int')
data.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0           0        0                  0          0        0            0   
1           0        0                  0          0        0            0   
2           0        0                  0          0        1            0   
3           0        0                  0          0        0            0   
4           0        0                  0          0        0            0   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0      0               0          0            0  ...       0               0   
1      0               0          0            0  ...       0               0   
2      0               0          0            0  ...       1               0   
3      0               0          0            0  ...       0               0   
4      0               0          0            0  ...       0               0   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0            0           0                 0                  0   
1            0           0                 0                  0   
2            0           0                 0                  0   
3            0           0                 0                  0   
4            0           0                 0                  0   

   whole wheat rice  yams  yogurt cake  zucchini  
0                 0     0            0         0  
1                 0     0            0         0  
2                 0     0            0         0  
3                 1     0            0         0  
4                 0     0            0         0  

[5 rows x 120 columns]

In [ ]:
data.isnull().sum() # checking for the null values

asparagus           0
almonds              0
antioxydant juice    0
asparagus            0
avocado              0
                    ..
whole wheat pasta    0
whole wheat rice     0
yams                 0
yogurt cake          0
zucchini             0
Length: 120, dtype: int64

In [ ]:
data.info() # information of the data after transformation

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Columns: 120 entries,  asparagus to zucchini
dtypes: int64(120)
memory usage: 6.9 MB


In [ ]:
data.describe() # describing the data

asparagus      almonds  antioxydant juice    asparagus     avocado  \
count  7500.000000  7500.000000        7500.000000  7500.000000  7500.00000   
mean      0.000133     0.020267           0.008800     0.004667     0.03320   
std       0.011547     0.140920           0.093401     0.068158     0.17917   
min       0.000000     0.000000           0.000000     0.000000     0.00000   
25%       0.000000     0.000000           0.000000     0.000000     0.00000   
50%       0.000000     0.000000           0.000000     0.000000     0.00000   
75%       0.000000     0.000000           0.000000     0.000000     0.00000   
max       1.000000     1.000000           1.000000     1.000000     1.00000   

       babies food        bacon  barbecue sauce    black tea  blueberries  \
count  7500.000000  7500.000000     7500.000000  7500.000000  7500.000000   
mean      0.004533     0.008667        0.010800     0.014267     0.009200   
std       0.067182     0.092697        0.103367     0.118596     0.095481   
min       0.000000     0.000000        0.000000     0.000000     0.000000   
25%       0.000000     0.000000        0.000000     0.000000     0.000000   
50%       0.000000     0.000000        0.000000     0.000000     0.000000   
75%       0.000000     0.000000        0.000000     0.000000     0.000000   
max       1.000000     1.000000        1.000000     1.000000     1.000000   

       ...       turkey  vegetables mix  water spray   white wine  \
count  ...  7500.000000     7500.000000  7500.000000  7500.000000   
mean   ...     0.062533        0.025600     0.000400     0.016533   
std    ...     0.242138        0.157949     0.019997     0.127523   
min    ...     0.000000        0.000000     0.000000     0.000000   
25%    ...     0.000000        0.000000     0.000000     0.000000   
50%    ...     0.000000        0.000000     0.000000     0.000000   
75%    ...     0.000000        0.000000     0.000000     0.000000   
max    ...     1.000000        1.000000     1.000000     1.000000   

       whole weat flour  whole wheat pasta  whole wheat rice         yams  \
count       7500.000000        7500.000000       7500.000000  7500.000000   
mean           0.009200           0.029467          0.058533     0.011333   
std            0.095481           0.169122          0.234765     0.105860   
min            0.000000           0.000000          0.000000     0.000000   
25%            0.000000           0.000000          0.000000     0.000000   
50%            0.000000           0.000000          0.000000     0.000000   
75%            0.000000           0.000000          0.000000     0.000000   
max            1.000000           1.000000          1.000000     1.000000   

       yogurt cake     zucchini  
count  7500.000000  7500.000000  
mean      0.027333     0.009467  
std       0.163064     0.096842  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       0.000000     0.000000  
max       1.000000     1.000000  

[8 rows x 120 columns]

## **Applying Aprior algorithm**

In [ ]:
frequent_itemsets = apriori(data, min_support=0.01, use_colnames=True)
frequent_itemsets

/usr/local/lib/python3.11/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                 itemsets
0    0.020267                                (almonds)
1    0.033200                                (avocado)
2    0.010800                         (barbecue sauce)
3    0.014267                              (black tea)
4    0.011467                             (body spray)
..        ...                                      ...
254  0.011067       (milk, ground beef, mineral water)
255  0.017067  (ground beef, spaghetti, mineral water)
256  0.015733         (milk, spaghetti, mineral water)
257  0.010267    (olive oil, spaghetti, mineral water)
258  0.011467     (pancakes, spaghetti, mineral water)

[259 rows x 2 columns]

#**Applying Association rule**

In [ ]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1.2,num_itemsets=3)
rules

antecedents                 consequents  \
0                     (avocado)             (mineral water)   
1               (mineral water)                   (avocado)   
2                        (cake)                   (burgers)   
3                     (burgers)                      (cake)   
4                        (eggs)                   (burgers)   
..                          ...                         ...   
345   (pancakes, mineral water)                 (spaghetti)   
346  (spaghetti, mineral water)                  (pancakes)   
347                  (pancakes)  (spaghetti, mineral water)   
348                 (spaghetti)   (pancakes, mineral water)   
349             (mineral water)       (pancakes, spaghetti)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.033200            0.238267  0.011467    0.345382  1.449559   
1              0.238267            0.033200  0.011467    0.048125  1.449559   
2              0.081067            0.087200  0.011467    0.141447  1.622103   
3              0.087200            0.081067  0.011467    0.131498  1.622103   
4              0.179733            0.087200  0.028800    0.160237  1.837585   
..                  ...                 ...       ...         ...       ...   
345            0.033733            0.174133  0.011467    0.339921  1.952073   
346            0.059733            0.095067  0.011467    0.191964  2.019260   
347            0.095067            0.059733  0.011467    0.120617  2.019260   
348            0.174133            0.033733  0.011467    0.065850  1.952073   
349            0.238267            0.025200  0.011467    0.048125  1.909736   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
0                 1.0  0.003556    1.163629       0.320785  0.044103   
1                 1.0  0.003556    1.015680       0.407144  0.044103   
2                 1.0  0.004398    1.063185       0.417349  0.073129   
3                 1.0  0.004398    1.058068       0.420154  0.073129   
4                 1.0  0.013127    1.086974       0.555682  0.120941   
..                ...       ...         ...            ...       ...   
345               1.0  0.005593    1.251163       0.504751  0.058384   
346               1.0  0.005788    1.119917       0.536836  0.080000   
347               1.0  0.005788    1.069235       0.557797  0.080000   
348               1.0  0.005593    1.034381       0.590560  0.058384   
349               1.0  0.005462    1.024084       0.625373  0.045503   

     certainty  kulczynski  
0     0.140620    0.196753  
1     0.015438    0.196753  
2     0.059430    0.136473  
3     0.054881    0.136473  
4     0.080015    0.245256  
..         ...         ...  
345   0.200744    0.202885  
346   0.107077    0.156291  
347   0.064752    0.156291  
348   0.033238    0.202885  
349   0.023518    0.251576  

[350 rows x 14 columns]

In [ ]:
rules.sort_values('lift',ascending = False)[0:20]

antecedents                         consequents  \
160                       (ground beef)                     (herb & pepper)   
161                     (herb & pepper)                       (ground beef)   
329                       (ground beef)          (spaghetti, mineral water)   
328          (spaghetti, mineral water)                       (ground beef)   
341                         (olive oil)          (spaghetti, mineral water)   
340          (spaghetti, mineral water)                         (olive oil)   
145                 (frozen vegetables)                          (tomatoes)   
144                          (tomatoes)                 (frozen vegetables)   
140                            (shrimp)                 (frozen vegetables)   
141                 (frozen vegetables)                            (shrimp)   
312                 (frozen vegetables)               (milk, mineral water)   
309               (milk, mineral water)                 (frozen vegetables)   
330                         (spaghetti)        (ground beef, mineral water)   
327        (ground beef, mineral water)                         (spaghetti)   
310  (frozen vegetables, mineral water)                              (milk)   
311                              (milk)  (frozen vegetables, mineral water)   
321               (milk, mineral water)                       (ground beef)   
324                       (ground beef)               (milk, mineral water)   
196                              (milk)                              (soup)   
197                              (soup)                              (milk)   

     antecedent support  consequent support   support  confidence      lift  \
160            0.098267            0.049467  0.016000    0.162822  3.291555   
161            0.049467            0.098267  0.016000    0.323450  3.291555   
329            0.098267            0.059733  0.017067    0.173677  2.907540   
328            0.059733            0.098267  0.017067    0.285714  2.907540   
341            0.065733            0.059733  0.010267    0.156187  2.614731   
340            0.059733            0.065733  0.010267    0.171875  2.614731   
145            0.095333            0.068400  0.016133    0.169231  2.474134   
144            0.068400            0.095333  0.016133    0.235867  2.474134   
140            0.071333            0.095333  0.016667    0.233645  2.450820   
141            0.095333            0.071333  0.016667    0.174825  2.450820   
312            0.095333            0.048000  0.011067    0.116084  2.418415   
309            0.048000            0.095333  0.011067    0.230556  2.418415   
330            0.174133            0.040933  0.017067    0.098009  2.394361   
327            0.040933            0.174133  0.017067    0.416938  2.394361   
310            0.035733            0.129600  0.011067    0.309701  2.389672   
311            0.129600            0.035733  0.011067    0.085391  2.389672   
321            0.048000            0.098267  0.011067    0.230556  2.346223   
324            0.098267            0.048000  0.011067    0.112619  2.346223   
196            0.129600            0.050533  0.015200    0.117284  2.320923   
197            0.050533            0.129600  0.015200    0.300792  2.320923   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
160               1.0  0.011139    1.135402       0.772060  0.121457   
161               1.0  0.011139    1.332841       0.732423  0.121457   
329               1.0  0.011197    1.137893       0.727562  0.121097   
328               1.0  0.011197    1.262427       0.697745  0.121097   
341               1.0  0.006340    1.114306       0.661001  0.089120   
340               1.0  0.006340    1.128171       0.656783  0.089120   
145               1.0  0.009613    1.121370       0.658605  0.109304   
144               1.0  0.009613    1.183913       0.639564  0.109304   
140               1.0  0.009866    1.180480       0.637444  

In [ ]:
rules[rules.confidence>=0.4]

antecedents      consequents  antecedent support  \
164                   (ground beef)  (mineral water)            0.098267   
206                     (olive oil)  (mineral water)            0.065733   
216                          (soup)  (mineral water)            0.050533   
250               (eggs, chocolate)  (mineral water)            0.033200   
262        (ground beef, chocolate)  (mineral water)            0.023067   
268               (milk, chocolate)  (mineral water)            0.032133   
280          (chocolate, spaghetti)  (mineral water)            0.039200   
286             (eggs, ground beef)  (mineral water)            0.020000   
292                    (milk, eggs)  (mineral water)            0.030800   
308       (milk, frozen vegetables)  (mineral water)            0.023600   
314  (frozen vegetables, spaghetti)  (mineral water)            0.027867   
320             (milk, ground beef)  (mineral water)            0.022000   
326        (ground beef, spaghetti)  (mineral water)            0.039200   
327    (ground beef, mineral water)      (spaghetti)            0.040933   
332               (milk, spaghetti)  (mineral water)            0.035467   
338          (olive oil, spaghetti)  (mineral water)            0.022933   
344           (pancakes, spaghetti)  (mineral water)            0.025200   

     consequent support   support  confidence      lift  representativity  \
164            0.238267  0.040933    0.416554  1.748266               1.0   
206            0.238267  0.027467    0.417850  1.753707               1.0   
216            0.238267  0.023067    0.456464  1.915771               1.0   
250            0.238267  0.013467    0.405622  1.702389               1.0   
262            0.238267  0.010933    0.473988  1.989319               1.0   
268            0.238267  0.014000    0.435685  1.828559               1.0   
280            0.238267  0.015867    0.404762  1.698777               1.0   
286            0.238267  0.010133    0.506667  2.126469               1.0   
292            0.238267  0.013067    0.424242  1.780536               1.0   
308            0.238267  0.011067    0.468927  1.968075               1.0   
314            0.238267  0.012000    0.430622  1.807311               1.0   
320            0.238267  0.011067    0.503030  2.111207               1.0   
326            0.238267  0.017067    0.435374  1.827256               1.0   
327            0.174133  0.017067    0.416938  2.394361               1.0   
332            0.238267  0.015733    0.443609  1.861817               1.0   
338            0.238267  0.010267    0.447674  1.878880               1.0   
344            0.238267  0.011467    0.455026  1.909736               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
164  0.017520    1.305576       0.474647  0.138475   0.234054    0.294175  
206  0.011805    1.308483       0.460018  0.099325   0.235756    0.266563  
216  0.011026    1.401441       0.503458  0.086804   0.286449    0.276637  
250  0.005556    1.281565       0.426758  0.052196   0.219704    0.231071  
262  0.005437    1.448130       0.509058  0.043663   0.309454    0.259938  
268  0.006344    1.349836       0.468165  0.054602   0.259169    0.247221  
280  0.006527    1.279712       0.428124  0.060652   0.218574    0.235677  
286  0.005368    1.544054       0.540548  0.040838   0.352354    0.274598  
292  0.005728    1.323011       0.452302  0.051042   0.244148    0.239541  
308  0.005444    1.434328       0.503778  0.044125   0.302809    0.257687  
314  0.005360    1.337834       0.459496  0.047219   0.252523    0.240493  
320  0.005825    1.532756       0.538177  0.044409   0.347580    0.274738  
326  0.007727    1.349094       0.471202  0.065540   0.258762    0.253501  
327  0.009939    1.416431       0.607207  0.086195   0.294000    0.257474  
332  0.007283    1.369061       0.479911  0.060982   0.269572    0.254821  
338  0.004802    1.379138       0.478747  0.040914  

# Analysis of Generated Rules:

### Frequent Combinations Involving "Mineral Water":

All the filtered rules involve "mineral water" as either the antecedent or the consequent. This indicates that "mineral water" is frequently purchased alongside other products in the dataset.

Some key combinations include:
- **Ground beef → Mineral water** (Lift: 1.748)
- **Olive oil → Mineral water** (Lift: 1.753)
- **Chocolate, eggs → Mineral water** (Lift: 1.702)
- **Eggs, ground beef → Mineral water** (Lift: 2.126)

### Strong Associations (Lift > 1.7):

The rules with the highest lift values, such as:
- **Eggs, ground beef → Mineral water** (Lift: 2.126)
- **Chocolate, ground beef → Mineral water** (Lift: 1.989)
- **Spaghetti, ground beef → Mineral water** (Lift: 1.827)

These combinations suggest that there are certain product pairs that have a notably high likelihood of being purchased together with mineral water, indicating a preference for these combinations.

### Confidence:

The confidence values range between 40% and 50%, meaning that, in these rules, there’s a 40%–50% chance that if a customer buys one of the products (such as "ground beef," "eggs," or "chocolate"), they will also buy "mineral water."

Confidence values above 0.4 suggest that these rules are strong enough to indicate a potential trend in customer behavior.

### Product Combinations:

- Multiple combinations involving **chocolate** appear, such as "chocolate, eggs" and "chocolate, milk," which indicates that customers who purchase chocolate are likely to buy mineral water as well.
- **Spaghetti** appears in several combinations, such as "spaghetti, frozen vegetables" and "spaghetti, ground beef," showing that spaghetti is often purchased with products like mineral water.
---

# Insights into Customer Purchasing Behavior:

### Mineral Water as a Common Purchase:

The fact that mineral water consistently appears in the rules suggests it is a staple product for many customers. It is likely to be purchased regularly alongside other food items.

### Pairing Preferences:

Customers seem to pair mineral water with more substantial items like **ground beef**, **eggs**, and **chocolate**. This could imply that when customers purchase more substantial or frequent grocery items (e.g., meat or snacks), they are also buying a refreshing beverage like mineral water.

### Combinations for Meal Preparation:

The presence of **spaghetti**, **chocolate**, and **eggs** in combinations suggests that customers might be shopping for complete meal solutions (spaghetti with other items) while also purchasing mineral water as an accompanying beverage.

### Promotional Opportunities:

Retailers could create bundle promotions around these frequent co-purchases, such as offering discounts when purchasing **mineral water** with **ground beef**, **eggs**, or **spaghetti**. This might increase sales by encouraging customers to buy these frequently-paired items together.

### Behavioral Segmentation:

The rules show that certain product types (e.g., chocolate, eggs, ground beef) are consistently purchased together with mineral water, which can help in segmenting customers based on their preferences. For instance, a segment of customers who regularly buy **chocolate** and **eggs** could be targeted with personalized promotions involving mineral water.

## 1. Support (Frequency of an Itemset)

**Definition:** Support measures how frequently an item or itemset appears in all transactions.

**Why is it needed?**
- Helps filter out infrequent items that are not useful.
- Ensures we only consider items that appear frequently enough to be relevant.

---

## 2. Confidence (Reliability of the Rule)

**Definition:** Confidence measures how often an item **B** is bought when item **A** is also bought.

**Why is it needed?**
- Helps identify strong association rules.
- Higher confidence means the rule is more trustworthy.


---

## 3. Lift (Strength of Association)

**Definition:** Lift measures how much more likely item **B** is bought when **A** is bought, compared to buying **B** independently.

**Why is it needed?**
- **Lift > 1** → A and B are positively correlated (buying A increases chances of buying B).
- **Lift = 1** → A and B are independent (no relationship).
- **Lift < 1** → A and B are negatively correlated (buying A reduces chances of buying B).